README

                                  ParaGUI

                       Crossplatform widgetset for SDL
		          Version 1.1.6 - UNSTABLE


http://www.paragui.org


This library is a high-level crossplatform application framework and GUI
library. It is completely based on the Simple DirectMedia Layer (SDL).
It can be compiled on various platforms (Linux, Win32, BeOS, MacOS, ...)
ParaGUI is targeted on crossplatform multimedia applications and
embedded devices operating on framebuffer displays.


Highlights of the library:

  * published under LGPL 

  * straight forward C++ class-design (see the API documentation) 

  * XML configuration of the new style engine 
  
  * XML widget layout loader
  
  * transparent useage of compressed file archives (with PhysicsFS)

  * asynchronous messagehandling (messageposts between objects) 

  * multithreaded 

  * support for non-modal overlapping widgets
    (windows inside of ParaGUI)

  * highly customizable widgets
    (background gradients, background images, transparency, colors, fonts) 

  * many standard widgets already implemented
    (buttons, labels, scrollbars, progressbars, windows ...) 

  * create your own widgets
    (subclass an existing widget) 

  * Python support

  * you can help us to make the library even better

  * signal / slot mechanism


This library is distributed under GNU LGPL version 2, which can be
found in the file "COPYING". 

To learn more about ParaGUI checkout the homepage, see the API
documentation and subscribe to the mailing list.

Here some rules that are a *must follow* for all contributors:


1. code in C++ - if there are standard C++ mechanisms that implement
   something (a list, a container, a buffer, a string etc.) then they 
   MUST be used unless some other interfaced code requires other techniques. 
   After all it's a C++ project

2. major changes that may break the functionality of ParaGUI and 
   the applications must be discussed on the devel list and _planned_
   before comitting them. The author of such changes should take every care 
   to implement the changes so that after commit the library is functional 
   in all its basic functionality.

3. If some changes require long-term development _and_ they break the
   functionality of the library, the developer should request for a 
   development branch of the current tree.


Have fun 


Alexander Pipelka
pipelka@teleweb.at


debian/changelog
libparagui1.1 (1.1.11-2) jaunty; urgency=low

  * fixed shlibs version

 -- Alexander Pipelka <pipelka@celeborn>  Wed, 09 Dec 2009 15:13:00 +0100

libparagui1.1 (1.1.11-1) jaunty; urgency=low

  * version 1.1.11

 -- Alexander Pipelka <pipelka@celeborn>  Wed, 09 Dec 2009 13:51:31 +0100

libparagui1.1 (1.1.10-2) jaunty; urgency=low

  * fixed shlibs version

 -- Alexander Pipelka <pipelka@celeborn>  Mon, 13 Jul 2009 14:25:49 +0200

libparagui1.1 (1.1.10-1) jaunty; urgency=low

  * added vnc server support
  * fixed the long staning config.h issue (no more PACKAGE already defined)
  * copyright header updates

 -- Alexander Pipelka <pipelka@celeborn>  Thu, 04 Jun 2009 12:12:19 +0200

libparagui1.1 (1.1.9-19ubuntu1) jaunty; urgency=low

  * fixed shlib version (debian package)

 -- Alexander Pipelka <pipelka@celeborn>  Thu, 07 May 2009 15:28:45 +0200

libparagui1.1 (1.1.9-19) jaunty; urgency=low

  * removed all the ext/hash map stuff

 -- Alexander Pipelka <pipelka@celeborn>  Wed, 06 May 2009 16:10:39 +0200

libparagui1.1 (1.1.9-18) feisty; urgency=low

  * updated stretch functions

 -- Alexander Pipelka <pipelka@celeborn>  Wed, 25 Jul 2007 16:15:00 +0200

libparagui1.1 (1.1.9-17) unstable; urgency=low

  * fixed PW_Draw zoom function (LockSurface bug)

 -- Alexander Pipelka <pipelka@celeborn.bms-austria.com>  Thu, 22 Mar 2007 15:56:12 +0100

libparagui1.1 (1.1.9-16) unstable; urgency=low

  * replaced all sprintf's with snprintf

 -- Alexander Pipelka <pipelka@celeborn.bms-austria.com>  Mon,  4 Dec 2006 13:19:38 +0100

libparagui1.1 (1.1.9-15) unstable; urgency=low

  * added missing dependency (libtiff4-dev)

 -- Alexander Pipelka <pipelka@bms-austria.com>  Fri,  8 Sep 2006 13:08:15 +0200

libparagui1.1 (1.1.9-14) unstable; urgency=low

  * rebuild

 -- Alexander Pipelka <pipelka@bms-austria.com>  Wed,  6 Sep 2006 11:19:41 +0200

libparagui1.1 (1.1.9-13) unstable; urgency=low

  * added PG_WidgetList::InsertAfter
  * various smaller fixes

 -- Alexander Pipelka <pipelka@bms-austria.com>  Wed, 26 Apr 2006 11:40:57 +0200

libparagui1.1 (1.1.9-12) unstable; urgency=low

  * fixed debian rules file

 -- Alexander Pipelka <pipelka@bms-austria.com>  Tue, 22 Nov 2005 10:22:05 +0100

libparagui1.1 (1.1.9-11) unstable; urgency=low

  * fixed building from CVS

 -- Alexander Pipelka <pipelka@shuttle.pipelka.net>  Mon, 21 Nov 2005 23:09:19 +0100

libparagui1.1 (1.1.9-10) unstable; urgency=low

  * dependency fixes for sarge / etch pbuilder environment

 -- Alexander Pipelka <pipelka@shuttle.pipelka.net>  Sun, 20 Nov 2005 16:03:11 +0100

libparagui1.1 (1.1.9-9) unstable; urgency=low

  * added libsdl-image1.2 dependency

 -- Alexander Pipelka <pipelka@teleweb.at>  Fri,  8 Jul 2005 10:35:20 +0200

libparagui1.1 (1.1.9-8) unstable; urgency=low

  * disabled exceptions, rtti

 -- Alexander Pipelka <pipelka@teleweb.at>  Fri, 10 Jun 2005 09:12:42 +0200

libparagui1.1 (1.1.9-7) unstable; urgency=low

  * made the debian package compatible with sarge and ubuntu (hoary, breezy)
    fixed some GCC 4 compiler warnings

 -- Alexander Pipelka <pipelka@teleweb.at>  Tue, 31 May 2005 22:07:48 +0200

libparagui1.1 (1.1.9-6) unstable; urgency=low

  * added PG_EventSupplier

 -- Alexander Pipelka <pipelka@teleweb.at>  Mon, 16 May 2005 18:21:37 +0200

libparagui1.1 (1.1.9-5) unstable; urgency=low

  * fix libc6 dependency to make the library installable on debian (testing) and ubuntu (hoary)

 -- Alexander Pipelka <pipelka@teleweb.at>  Tue,  3 May 2005 15:33:29 +0200

libparagui1.1 (1.1.9-4) unstable; urgency=low

  * made PG_Widget::Show/Hide methods virtual 

 -- Alexander Pipelka <pipelka@teleweb.at>  Mon,  7 Feb 2005 14:15:43 +0100

libparagui1.1 (1.1.9-3) unstable; urgency=low

  * fixed duplicated libparagui.so 

 -- Alexander Pipelka <pipelka@teleweb.at>  Mon, 20 Dec 2004 11:19:16 +0100

libparagui1.1 (1.1.9-2) unstable; urgency=low

  * changed compiler flags (for non-debug builds), removed static libs from
    dev package

 -- Alexander Pipelka <pipelka@teleweb.at>  Sat, 18 Dec 2004 20:07:32 +0100

libparagui1.1 (1.1.9-1) unstable; urgency=low

  * changed version to 1.1.9

 -- Alexander Pipelka <pipelka@teleweb.at>  Wed, 17 Nov 2004 19:18:16 +0100

libparagui1.1 (1.1.8-1) unstable; urgency=low

  * Initial unstable package

 -- Alexander Pipelka <pipelka@teleweb.at>  Mon, 18 Oct 2004 19:21:22 +0100

Local variables:
mode: debian-changelog
End:


include/pglog.h
/*
    ParaGUI - crossplatform widgetset
    Copyright (C) 2000 - 2009 Alexander Pipelka

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public
    License along with this library; if not, write to the Free
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

    Alexander Pipelka
    pipelka@teleweb.at

    Last Update:      $Author: braindead $
    Update Date:      $Date: 2009/06/04 10:25:04 $
    Source File:      $Source: /sources/paragui/paragui/include/pglog.h,v $
    CVS/RCS Revision: $Revision: 1.3.6.2.2.7 $
    Status:           $State
*/

/**
	\file pglog.h
	ParaGUI logging facility.
	This header declares the PG_LogConsole namespace and all functions for log processing.
*/

#ifndef PGLOG_H
#define PGLOG_H

#include <stdarg.h>
#include "SDL.h"
#include "paragui.h"
#include "pglabel.h"

//! ParaGUI logging level
typedef enum {
    PG_LOG_NONE,  //!< log nothing
    PG_LOG_ERR, //!< only errors
    PG_LOG_WRN, //!< errors and warnings
    PG_LOG_MSG, //!< display errors, warnings and messages
    PG_LOG_DBG //!< log everything
} PG_LOG_LEVEL;

/**
	Logmethod: STDOUT.
	Send all log messages to stdout.
*/
#define	PG_LOGMTH_STDOUT 	1
/**
	Logmethod: STDERR.
	Send all log messages to stderr.
*/
#define	PG_LOGMTH_STDERR	2
/**
	Logmethod: CONSOLE.
	Send all log messages to the console window.
*/
#define	PG_LOGMTH_CONSOLE  	4

#ifdef __cplusplus

/**
	@short logging functions.

	These functions can be used for any purpose logging
*/
namespace PG_LogConsole {
/**
Log a message

Don't use this function directly, it is unconvenient and meant for
internal use only. Use The PG_Log functions instead if you want to
log something.

@param	id		the log level of the message
@param	Text	the message text *printf format string)
@param	ap		the variable list
*/
DECLSPEC void LogVA(PG_LOG_LEVEL id, const char *Text, va_list ap);

/**
Clears the list of log messages and removes the log window
*/
DECLSPEC void Done();

/**
Updates the log window

This function creates a log window if it doesn't exist and writes
all log messages to this window.
*/
DECLSPEC void Update();

/**
Set the minimum log level

@param	newlevel	the new minimum log level
All log messages with a log level smaller than newlevel will be
quietly discarded.
*/
DECLSPEC void SetLogLevel(PG_LOG_LEVEL newlevel);

/**
Show the log window
*/
DECLSPEC void Show();

/**
Hide the log window
*/
DECLSPEC void Hide();

/**
Toggle between log window shown/hidden
*/
DECLSPEC void Toggle();

/**
Set the log method

@param	method	determines where to log messages to
*/
DECLSPEC void SetMethod(int method);

/**
Get the log method

@return	current log method
*/
DECLSPEC int GetMethod();

/**
Set the key to toggle the console (default F12)

@param	key	the key to toggle the console
The key will be used by PG_Application to toggle the log console.
*/
DECLSPEC void SetConsoleKey(SDLKey key);

/**
Get the current key to toggle console

@return	current key to toggle the console
*/
DECLSPEC SDLKey GetConsoleKey();

/**
Set the title of the log console window

@param	title	the title of the window
@param	alignment	the alingment of the title
*/
DECLSPEC void SetTitle(const char* title, PG_Label::TextAlign alignment = PG_Label::CENTER);

/**
Set a limit for the amount of log lines (200 lines by default).

@param	max maximum amount of log lines.
*/
DECLSPEC void SetMaxLogLines(Uint32 max);
}

#endif

#ifdef __cplusplus
extern "C" {
#endif

	/**
	Log a message

	@param	id	the log level id of the message
	@param	Text	a printf format string, arbitrary arguments may follow
	*/
	DECLSPEC void PG_Log(PG_LOG_LEVEL id, const char *Text, ...);

	/**
	Log a message with log level "message"

	@param	fmt	printf format log string
	*/
	DECLSPEC void PG_LogMSG(const char *fmt, ...);

	/**
	Log a message with log level "message"

	@param	fmt	printf format log string
	@param  ap	argument list
	*/
	DECLSPEC void PG_LogMSGV(const char *fmt, va_list ap);

	/**
	Log a message with log level "error"

	@param	fmt	printf format string
	*/
	DECLSPEC void PG_LogERR(const char *fmt, ...);

	/**
	Log a message with log level "error"

	@param	fmt	printf format log string
	@param  ap	argument list
	*/
	DECLSPEC void PG_LogERRV(const char *fmt, va_list ap);

	/**
	Log a message with log level "warning"

	@param	fmt	printf format string
	*/
	DECLSPEC void PG_LogWRN(const char *fmt, ...);

	/**
	Log a message with log level "warning"

	@param	fmt	printf format log string
	@param  ap	argument list
	*/
	DECLSPEC void PG_LogWRNV(const char *fmt, va_list ap);

	/**
	Log a message with log level "debug"

	@param	fmt	printf format string
	*/
	DECLSPEC void PG_LogDBG(const char *fmt, ...);

	/**
	Log a message with log level "debug"

	@param	fmt	printf format log string
	@param  ap	argument list
	*/
	DECLSPEC void PG_LogDBGV(const char *fmt, va_list ap);

#ifdef __cplusplus
}
#endif

#endif // PGLOG_H

/*
 * Local Variables:
 * c-basic-offset: 8
 * End:
 */
